### ISSUE TITLE
Sentry no longer capturing request body

### ISSUE THREAD

**Issue by mfb** [getsentry/sentry-php#1135]
It appears that after #1119 the RequestIntegration is no longer capturing the request body.

As far as I can tell, there **is** a request body, i.e. getBody() returns e.g. a URL-encoded POST request, but the getSize() method returns zero.

**Comment by mfb** [comment]
It appears that the size is 0 until you read at least one byte of the body.

**Comment by mfb** [comment]
I don't actually grok what the original issue in #1118 was, so not sure how to fix.

**Comment by ste93cry** [comment]
>there **is** a request body, i.e. getBody() returns e.g. a URL-encoded POST request, but the getSize() method returns zero

The issue seems to be that `php://input` is not a file and thus calling `fstat` to get its size does not work. The Guzzle PSR-7 implementation wraps the stream into a `CachingStream` which computes the size as it reads from the underlying stream, but as you said until you start reading it will be `0`

>I don't actually grok what the original issue in #1118 was

The issue was that since `getSize()` was returning `0` all comparison with `$maxRequestBodySize` were failing and a very big request body could have been read leading to an OOM error. What I did not notice at that time is that the size is always `0` in that place

>so not sure how to fix

I think that we should be safe in reading the `Content-Length` HTTP header in place of using `getSize()`

**Comment by Jean85** [comment]
Wouldn't put a limit to reading the stream a better approach?

**Comment by mfb** [comment]
A limit on reading the stream seems safer, since there could be edge cases, like request is compressed (content length is not body size) or chunked transfer (no content length).

By the way, is there enough memory in the default config when handling fatal errors? 10240 bytes reserved, 10000 byte max content length. That leaves only 240 bytes left over.

**Comment by ste93cry** [comment]
>Wouldn't put a limit to reading the stream a better approach?

I don't understand what this has to do with the issue, can you explain a bit more what you meant?

**Comment by mfb** [comment]
Rather than `$requestBody->getContents()` you could read just part of it `$requestBody->read(MAX_CONTENT_LENGTH)` - presumably return null if the stream is not at end of file

**Comment by ste93cry** [comment]
Aren't we talking about a different issue? The original problem is that right now no request data is logged at all, regardless of the `max_request_body_size` option. If we want to discuss about limiting how much data we read later then we should open a separate issue, shouldn't we?

**Comment by mfb** [comment]
Hopefully both issues could be solved in the same code? Basically this issue could be a re-do of #1118 - given that running out of memory is worse than failing to log the request body.

(I actually found not logging the request body by default to be an improvement for data security, but of course it wasn't intended :)

**Comment by ste93cry** [comment]
I had a look at two SDK implementations, the one of Go and the one of Python and this is how they work:

|  | Go | Python |
|-|-|-|
| `$_SERVER['CONTENT_LENGTH'] > $maxRequestBodySize` | Do nothing | Set empty string annotated with the content length |
| `$_SERVER['CONTENT_LENGTH'] <= $maxRequestBodySize` | Read at most `$maxRequestBodySize` bytes | Read parsed body or fallback to empty string annotated with the content length |

Both implementation use the Content-Length header to get the size of the request body. In an ideal world we could take the best of both approaches which imho is to limit the stream and annotate the string with the content length, but since this feature is not supported in PHP we should decide what to do when the parsed body is empty: should we still try to read from the stream (even if limited) or should we simply not log anything?

**Comment by mfb** [comment]
What part of this can't be done in PHP - can't you annotate it with with the content length if it's not at end of file after reading $maxRequestBodySize bytes?

**Comment by ste93cry** [comment]
>What part of this can't be done in PHP

It's not that it cannot be done in PHP, but the Python SDK has some kind of feature that we currently don't have that I never saw before that basically works by annotation the values with some special "metadata" so that the server treats them differently. As far as I remember it's an implementation that somehow is tied with [Relay](https://getsentry.github.io/relay/relay/) I think

**Comment by mfb** [comment]
Neat. There must be some way to do this if it's just json data?

**Comment by ste93cry** [comment]
Indeed, it's just a format for the JSON value that is treated in a special way on the server, but I never investigated how to make it work to be honest. Anyway, I don't know which value it brings on the UI so I would leave this discussion for another day and just focus on the main question:

>we should decide what to do when the parsed body is empty: should we still try to read from the stream (even if limited) or should we simply not log anything?

We can add all the nice things after fixing this bug because actually it's important to release a patch that fixes the regression as a lot of apps out of here are not logging anymore the request data
